### PR TITLE
Bookstore: add reducers and actions

### PR DESCRIPTION
**Pull request by developertariq**
**- Added Redux Toolkit**
**- Used the ducks pattern for your Redux files**
**- Written your book's actions and reducer**
**- Writtenyour categories actions and reducer**
**- Configured the Redux Store**

### CODE REVIEW

File: src/redux/books/Books.js [hunk 1]
```diff
@@ -1,9 +1,35 @@
 import React, { useState } from 'react';
 import { v4 as uuidv4 } from 'uuid';
-import BookList from '../components/BookList';
-import AddBook from '../components/AddBook';
+import BookList from '../../components/BookList';
+import AddBook from '../../components/AddBook';
 
-const Books = () => {
+const ADD = 'ADD';
+const REMOVE = 'REMOVE';
```
Comment (rloterh) [inline src/redux/books/Books.js]: - [ ] With reference to the _duck pattern_, assigning values to _ADD_BOOK_ and _REMOVE_BOOK_ action types should be something like:
```ruby
       const ADD_BOOK = bookstore/books/ADD_BOOK;
```

and

```ruby
      const REMOVE_BOOK = bookstore/books/REMOVE_BOOK;
```

Kindly ensure to improve upon these aspects.

File: src/redux/categories/Categories.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+const STATUS = 'STATUS';
```
Comment (rloterh) [inline src/redux/categories/Categories.js]: - [ ] Similarly, assigning value to _CHECK_STATUS_ action type with respect to the _duck patterns approach_ should be something like below:
```ruby
      CHECK_STATUS = bookstore/categories/CHECK_STATUS;
```
 
Kindly ensure to improve upon this aspect.

File: src/redux/categories/Categories.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+export default function categories(state = initialState, action) {
+  switch (action.type) {
+    case STATUS:
+      return { ...state, categories: action.categories };
+    default:
+      return state;
+  }
+}
```
Comment (rloterh) [inline src/redux/categories/Categories.js]: - [ ] Case action.type is `CHECK_STATUS` the `categories` function should return a string _"Under construction"_. Please improve upon this as per this module's requirement

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Tariq, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦  🏃‍♂️
 
## Highlights 👍  
- Nice job so far!
- Professional ReadMe!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/redux/categories/Categories.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      return { ...state, categories: action.categories, status: 'Under construction' };
+    default:
+      return state;
+  }
+}
+
+export const Categories = () => (
+  <div className="container">
+    <button type="button">Check status</button>
+  </div>
```
Comment (BenMukebo) [inline src/redux/categories/Categories.js]: > * [ ]  Case action.type is `CHECK_STATUS` the `categories` function should return a string _"Under construction"_. Please improve upon this as per this module's requirement
 

- [ ] Great job so far, however as requested by the previous reviewer, Make sure the `Check status` working fine and return the `Under construction` message

File: src/redux/books/Books.js [hunk 5]
```diff
@@ -1,9 +1,35 @@
...
+const REMOVE_BOOK = 'bookstore/books/REMOVE_BOOK';
+
+export function addBook(book) {
+  return { type: ADD_BOOK, book };
+}
+
+export function removeBook(book) {
+  return { type: REMOVE_BOOK, book };
+}
+
```
Comment (BenMukebo) [inline src/redux/books/Books.js:16]: - [ ] Actually, I can create a book from the form but the remove button is no longer working, Kindly fix this 💯 👍🏻

**Review by BenMukebo — CHANGES_REQUESTED** [review]
Hi @developertariq, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight :clap: 
- No linter errors. :heavy_check_mark:
- GitHub flow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:


 ## Required Changes ♻️
_Check the comments under the review._


You have submitted a project, that Linters check are not running in the GitHub/action
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please Tag me on (@BenMukebo) when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/redux/books/Books.js (hunk 5 above)
Comment (developertariq) [reply src/redux/books/Books.js:16]: Hi @BenMukebo, 

Thanks for your feedback. There is no requirement for this. Will you please check again the requirements.

Thanks again

File: src/redux/categories/Categories.js (hunk 4 above)
Comment (developertariq) [reply src/redux/categories/Categories.js]: Hi @BenMukebo, 
Thanks for your feedback. Could you please help me to resolve this issue and check the requirements.

Thanks

File: src/redux/categories/Categories.js (hunk 2 above)
Comment (developertariq) [reply src/redux/categories/Categories.js]: Hi @rloterh, 
Thanks for your feedback. I am sharing somethink for you. If I am not wrong duck pattenrs is ---
The Ducks pattern is a way to organize the code for a feature in a Redux application. It involves grouping together the action types, action creators, and the reducer for a feature in a single file, typically called index.js or [featureName].js.

The idea behind the Ducks pattern is to keep the code for a feature together in one place, making it easier to understand and maintain. This pattern also promotes reusability of the code.

The file that follows the Ducks pattern should export the action types, action creators, and the reducer. The action types and action creators are used to describe and create the actions that are dispatched to the store, while the reducer is used to handle the state updates related to those actions.

The Ducks pattern is not a standard or a library, but a widely used pattern in the community that follows few conventions like:

- The file name should be in the plural form.
- The exported values should be the default export.
- The default export should be a combination of the reducer and the actions.
- The action types should be defined as constants and should be exported.

Overall, the Ducks pattern is a way to organize your code and make it more readable and maintainable.

Comment (rloterh) [reply src/redux/categories/Categories.js]: Thanks for sharing @developertariq 👍

File: src/redux/categories/Categories.js (hunk 4 above)
Comment (BenMukebo) [reply src/redux/categories/Categories.js]: - [ ] Hello @developertariq, As you can see on the screenshort 👇🏻 these are part of your project requiement I did not invite anything.

im2

**Review by nicupop729 — COMMENTED** [review]
Hello @developertariq 👋,

_This is an additional review as a result of the second opinion request._

Kudos for the effort put in place until now for this project :clap:

To progress to the next milestone, kindly make sure to implement the project requirements of this lesson and follow the guidance from the previous reviewers:
- please make sure to return an "Under Construction" string when the user clicks the "Check Status" button:
image
- also, kindly ensure that the item is deleted if the user clicks on the "Remove" button:
image
image

After implementing the project requirements, you can safely request another review from your dashboard!

Keep rocking 🚀!

**Comment by developertariq** [comment]
Hi @BenMukebo @nicupop729 
Thanks for your feedback. Sorry for your inconvenient. If I am not wrong, I have found this requirements in second day exercise.
[Bookstore: use Redux in React components.](https://github.com/microverseinc/curriculum-react-redux/blob/main/bookstore/use_redux_in_react_components.md). 

day2

Will you please help me to resolve this issue.

Thanks and regards.

**Review by nicupop729 — COMMENTED** [review]
Hello @developertariq  👋,

I apologise, you are right.
For this milestone, the implementation of the functionality should not be put in place, but only to set the actions and reducers for redux.

Your PR is ready to be approved! :clap:

As this was the only reason that your project was not approved it will be approved as soon as possible.
Please ask for another review in your Dashboard and the next Code Reviewer will mark it as approve right away :)

Keep rocking 🚀!

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi @developertariq, 

### Good job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good communication with code reviewers
:heavy_check_mark: Clean and organized code

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can get to me on slack  using this handle @mihndim.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**